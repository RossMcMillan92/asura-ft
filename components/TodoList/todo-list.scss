@import '../../styles/settings/breakpoints.scss';
@import '../../styles/settings/pallete.scss';

.todo-list {
  width: 100%;
  max-width: 500px;
  background-color: #fff;
  box-shadow: 0 1px 3px 0px rgba(0, 0, 0, 0.15),
    0 5px 32px -14px rgba(0, 0, 0, 0.35);
  border-radius: 7px;
  overflow: hidden;
}

.todo-list__title {
  font-size: 24px;
  font-weight: 300;
  margin: 25px 0;
  color: $base-text-color;
  text-align: center;

  @media (min-width: $breakpoint-tablet) {
    font-size: 30px;
  }

  @media (min-width: $breakpoint-desktop) {
    font-size: 38px;
  }
}

.todo-list__list {
  list-style: none;
  margin: 0;
  padding: 0;
}

.todo-list__entry-section {
  background-color: #f7f7f7;
  padding: 15px;

  @media (min-width: $breakpoint-tablet) {
    padding: 25px;
  }
}

.todo-list__empty-message {
  height: 62px;
  display: flex;
  align-content: center;
  justify-content: center;

  @media (min-width: $breakpoint-tablet) {
    height: 92px;
  }
}

.todo-list__empty-message-text {
  font-size: 16px;
  color: $base-text-color;
  animation: fade-in 1s ease-in-out;

  @media (min-width: $breakpoint-tablet) {
    font-size: 19px;
  }
}

@keyframes fade-in {
  0% {
    opacity: 0;
    transform: translateY(10%);
  }
  100% {
    opacity: 1;
    transform: translateY(0);
  }
}
