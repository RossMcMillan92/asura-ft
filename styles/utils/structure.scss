.u-flex-row {
  display: flex !important;
  flex-direction: row !important;
}

.u-flex-column {
  display: flex !important;
  flex-direction: column !important;
}

.u-flex-justify-content-center {
  justify-content: center !important;
}

.u-flex-grow-1 {
  flex-grow: 1 !important;
}

.u-padding-15 {
  padding: 15px !important;
}

.u-padding-20 {
  padding: 20px !important;
}

.u-pull-up-170 {
  margin-top: -170px !important;
}

.u-margin-right-20 {
  margin-right: 20px !important;
}

.u-visually-hidden {
  position: absolute !important;
  clip: rect(1px, 1px, 1px, 1px) !important;
  padding: 0 !important;
  border: 0 !important;
  height: 1px !important;
  width: 1px !important;
  overflow: hidden !important;
}
