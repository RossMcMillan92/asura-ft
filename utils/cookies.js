import { getOr, pipe } from 'lodash/fp'
import { parseCookies, setCookie } from 'nookies'

const TODO_COOKIE_KEY = 'todos'

export const getTodosFromCookies = pipe(
  parseCookies,
  getOr('null', TODO_COOKIE_KEY),
  JSON.parse
)

export const setTodosCookie = pipe(
  JSON.stringify,
  todos =>
    setCookie({}, TODO_COOKIE_KEY, todos, {
      maxAge: 30 * 24 * 60 * 60,
      path: '/'
    })
)
