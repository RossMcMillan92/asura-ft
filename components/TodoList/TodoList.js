import './todo-list.scss'

import { pipe, remove, set, tap } from 'lodash/fp'

import AddTodoForm from '../AddTodoForm/AddTodoForm'
import PropTypes from 'prop-types'
import React from 'react'
import TodoItem from '../TodoItem/TodoItem'
import { setTodosCookie } from '../../utils/cookies'

const TodoList = ({ storedTodos }) => {
  const [todos, setTodos] = React.useState(storedTodos)
  const [announcerMessage, setAnnouncerMessage] = React.useState('')
  const headerRef = React.useRef()
  const headingId = 'todo-list-heading'

  const setAndPersistTodos = pipe(
    tap(setTodos),
    tap(setTodosCookie)
  )

  const onTodoCheckboxClick = (todo, index) => {
    setAndPersistTodos(set([index, 'checked'], !todo.checked, todos))
  }

  const onAddTodo = newTodo => {
    const todoAlreadyExists = todos.some(todo => todo.description === newTodo)
    if (todoAlreadyExists) return
    setAndPersistTodos(todos.concat({ description: newTodo, checked: false }))
    setAnnouncerMessage(`${newTodo} added.`)
  }

  const onRemoveTodo = index => {
    setAndPersistTodos(remove(todo => todo === todos[index], todos))
    setAnnouncerMessage(`${todos[index].description} removed.`)
    headerRef.current.focus()
  }

  return (
    <section aria-labelledby={headingId} className="todo-list">
      <h1
        id={headingId}
        className="todo-list__title"
        tabIndex="-1"
        ref={headerRef}
      >
        To Do
      </h1>
      {todos.length !== 0 && (
        <ul className="todo-list__list">
          {todos.map((todo, index) => (
            <TodoItem
              key={todo.description}
              id={`todo-${index}`}
              name="todo[]"
              label={todo.description}
              value={todo.description}
              checked={todo.checked}
              onChange={() => onTodoCheckboxClick(todo, index)}
              onDelete={() => onRemoveTodo(index)}
            />
          ))}
        </ul>
      )}
      {todos.length === 0 && (
        <div className="todo-list__empty-message">
          <p className="todo-list__empty-message-text">You&apos;re all done!</p>
        </div>
      )}
      <div className="todo-list__entry-section">
        <AddTodoForm onSubmit={onAddTodo} />
      </div>
      <div
        role="status"
        aria-live="polite"
        className="todo-list__announcer u-visually-hidden"
      >
        {announcerMessage}
      </div>
    </section>
  )
}

TodoList.propTypes = {
  storedTodos: PropTypes.arrayOf(
    PropTypes.shape({
      description: PropTypes.string.isRequired,
      checked: PropTypes.bool.isRequired
    })
  ).isRequired
}

export default TodoList
