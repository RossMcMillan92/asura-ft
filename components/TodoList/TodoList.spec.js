import { mount, shallow } from 'enzyme'

import React from 'react'
import TodoList from './TodoList'
import { act } from 'react-dom/test-utils'
import { setTodosCookie } from '../../utils/cookies'

jest.mock('../../utils/cookies', () => ({
  setTodosCookie: jest.fn()
}))

const props = {
  storedTodos: [
    { description: 'Task 1', checked: false },
    { description: 'Task 2', checked: true }
  ]
}
const getWrapper = additionalProps =>
  mount(<TodoList {...props} {...additionalProps} />)
const getShallowWrapper = additionalProps =>
  shallow(<TodoList {...props} {...additionalProps} />)

describe('TodoList', () => {
  it('matches the snapshot', () => {
    const wrapper = getShallowWrapper()
    expect(wrapper).toMatchSnapshot()
  })

  describe('when there are stored todos', () => {
    it('lists the todos in a list element', () => {
      const wrapper = getWrapper()
      const list = wrapper.find('ul')
      const listItems = wrapper.find('ul li')

      expect(list.exists()).toBe(true)
      expect(listItems.length).toBe(props.storedTodos.length)
      expect(listItems.at(0).text()).toContain(props.storedTodos[0].description)
      expect(
        listItems
          .at(0)
          .find('[type="checkbox"]')
          .hostNodes()
          .prop('checked')
      ).toBe(props.storedTodos[0].checked)
      expect(listItems.at(1).text()).toContain(props.storedTodos[1].description)
      expect(
        listItems
          .at(1)
          .find('[type="checkbox"]')
          .hostNodes()
          .prop('checked')
      ).toBe(props.storedTodos[1].checked)
    })
  })

  describe('when the rendered Add Todo form is submitted', () => {
    const wrapper = getWrapper()
    const newTodoValue = 'To do task...'

    beforeAll(() => {
      act(() => {
        wrapper
          .find('#todo-entry')
          .hostNodes()
          .simulate('change', { target: { value: newTodoValue } })
      })
      wrapper.update()

      act(() => {
        wrapper.find('form').simulate('submit')
      })
      wrapper.update()
    })

    it('renders a new unchecked todo item in the list', () => {
      const listItems = wrapper.find('ul li')

      expect(listItems.length).toBe(props.storedTodos.length + 1)
      expect(listItems.at(0).text()).toContain(props.storedTodos[0].description)
      expect(listItems.at(1).text()).toContain(props.storedTodos[1].description)
      expect(listItems.at(2).text()).toContain(newTodoValue)
      expect(
        listItems
          .at(2)
          .find('[type="checkbox"]')
          .hostNodes()
          .prop('checked')
      ).toBe(false)
    })

    it('stores the todos in the cookies', () => {
      expect(setTodosCookie).toBeCalledWith([
        ...props.storedTodos,
        { description: newTodoValue, checked: false }
      ])
    })

    describe('when the same todo is added twice', () => {
      beforeAll(() => {
        act(() => {
          wrapper
            .find('#todo-entry')
            .hostNodes()
            .simulate('change', { target: { value: newTodoValue } })
        })
        wrapper.update()

        act(() => {
          wrapper.find('form').simulate('submit')
        })
        wrapper.update()
      })

      it('does not render the second one', () => {
        const listItems = wrapper.find('ul li')
        expect(listItems.length).toBe(props.storedTodos.length + 1)
      })
    })
  })

  describe("when a todo's delete button is clicked", () => {
    const wrapper = getWrapper()
    const todoItemToBeDeleted = wrapper.find('ul li').at(0)

    beforeAll(() => {
      expect(wrapper.find('ul li').hostNodes().length).toBe(
        props.storedTodos.length
      )

      act(() => {
        todoItemToBeDeleted
          .find('[data-testid="delete-button"]')
          .hostNodes()
          .simulate('click')
      })
      wrapper.update()
    })

    it('removes the todo item from the list', () => {
      expect(wrapper.find('ul li').hostNodes().length).toBe(
        props.storedTodos.length - 1
      )
      expect(wrapper.find('ul li').text()).not.toContain(
        props.storedTodos[0].description
      )
    })

    it('stores the new todo list in the cookies', () => {
      expect(setTodosCookie).toBeCalledWith(
        props.storedTodos.slice(1, props.storedTodos.length)
      )
    })
  })
})
