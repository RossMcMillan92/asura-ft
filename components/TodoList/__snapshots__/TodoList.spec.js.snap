// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList matches the snapshot 1`] = `
<section
  aria-labelledby="todo-list-heading"
  className="todo-list"
>
  <h1
    className="todo-list__title"
    id="todo-list-heading"
    tabIndex="-1"
  >
    To Do
  </h1>
  <ul
    className="todo-list__list"
  >
    <TodoItem
      checked={false}
      id="todo-0"
      key="Task 1"
      label="Task 1"
      name="todo[]"
      onChange={[Function]}
      onDelete={[Function]}
      value="Task 1"
    />
    <TodoItem
      checked={true}
      id="todo-1"
      key="Task 2"
      label="Task 2"
      name="todo[]"
      onChange={[Function]}
      onDelete={[Function]}
      value="Task 2"
    />
  </ul>
  <div
    className="todo-list__entry-section"
  >
    <AddTodoForm
      onSubmit={[Function]}
    />
  </div>
  <div
    aria-live="polite"
    className="todo-list__announcer u-visually-hidden"
    role="status"
  />
</section>
`;
